docs/spec.md

```diff
@@ -160,10 +163,56 @@ following constraints:
 * (C12) If `quantization_dimension` is not empty, then
   `0 <= quantization_dimension`.
 
-At the moment, `QuantizationScale` is a floating-point constant, but there is
-strong interest in integer-based scales, represented with multipliers and
-shifts. We are planning to explore this in the near future
-([#1404](https://github.com/openxla/stablehlo/issues/1404)).
+In order to allow operation using only integer arithmetic, the floating-point
+`scale`, `S`, is realized using integer `multipler`, `M` and `shift` value,
+`n >= 0`, such that `round_nearest_integer(S) = (M + round)* 2^-n`, where
+`round` is added to recover the precision loss while deriving the values of
+`M` and `n` from `S`.
+
+The following demonstrates, using C++ code, a possible implementation of
+deriving the integer parameters `M` and `n` of type `i32` from a
+floating-point scale `S` of type `f64`:
+
+```c++
+    int32_t shift;
+    double mantissa = std::frexp(S, &shift);
+    auto shiftedMantissa = static_cast<int64_t>(std::round(mantissa * (int64_t(1) << 31)));
+
+    // Ensure that shiftedMantissa is within limits of 32-bit integer type.
+    if (shiftedMantissa == (int64_t(1) << 31)) {
+        shiftedMantissa /= 2;
+        shift++;
+    }
+
+    int64_t adjustedMantissa = shiftedMantissa;
+
+    // For small `S` value, flush `adjustedMantissa` to zero.
+    if (shift < -31) {
+        shift = 0;
+        adjustedMantissa = 0;
+    }
+
+    // Saturate `adjustedMantissa`, if shift > 30,
+    if (shift > 30) {
+        shift = 30;
+        adjustedMantissa = (1LL << 31) - 1;
+    }
+
+    n = (-shift) + 31;
+    M = static_cast<int32_t>(adjustedMantissa);
+```
+
+Using the above implementation, the following C++ code demonstrates the
+relationship between the floating-point scale `S` and the integer parameters `M`
+and `n`.
+
+```c++
+    int64_t round = static_cast<int64_t>(1) << (n - 1);
+    static_cast<int32_t>(std::floor(S + 0.5)) = clamp(
```

Comment:
Similar to the comment on line 168, I think both sides become zero in the case of small S like 0.25 or lower. The left hand side could be changed to floor(S*(1<<n) + 0.5) and right hand size to M.

Reply:
Also updated the code here. PTAL.

Reply:
Thanks - one new comment below on the updated line.